File: example-android-project/app/src/main/java/io/github/gofaith/faithdroid/FViews/FPermission.java
package io.github.gofaith.faithdroid.FViews;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.support.v4.app.ActivityCompat;
import android.util.Log;

import io.github.gofaith.faithdroid.UI.AttrGettable;
import io.github.gofaith.faithdroid.UI.AttrSettable;
import io.github.gofaith.faithdroid.UI.UIController;

import static android.content.ContentValues.TAG;

public class FPermission  {
    public static String getAttr(UIController parentController,String attr) {
        String[] strs = attr.split(":");
        String method=strs[0],arg=strs[1];
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            switch (method) {
                case "CheckSelfPermission":
                    return parentController.activity.checkSelfPermission(arg) == PackageManager.PERMISSION_GRANTED ? "true" : "false";
            }
        }
        return null;
    }

    public static void setAttr(UIController parentController,String attr, String value) {
        if (value == null) {
            return;
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            switch (attr) {
                case "RequestPermissions":
                    String[] strs = value.split(":");
                    // fnId=strs[0] , permissions=strs[1].split("|") , requestCode=strs[2]
                    Integer requestCode = Integer.valueOf(strs[2]);
                    String[] ps=strs[1].split("#");
                    parentController.onPermissionResults.put(requestCode, strs[0]);
                    Log.d(TAG, "setAttr: "+value);
                    ActivityCompat.requestPermissions(
                            parentController.activity,
                            ps,
                            requestCode);
                    break;
            }
        }
    }
}

File: android/app/src/main/java/io/github/stevenzack/faithdroidapp/FViews/FPermission.kt
package io.github.stevenzack.faithdroidapp.FViews

import android.content.ContentValues.TAG
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.core.app.ActivityCompat
import io.github.stevenzack.faithdroidapp.UI.UIController


object FPermission {
    fun getAttr(parentController: UIController, attr: String): String? {
        val strs = attr.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val method = strs[0]
        val arg = strs[1]
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            when (method) {
                "CheckSelfPermission" -> return if (parentController.activity.checkSelfPermission(arg) === PackageManager.PERMISSION_GRANTED) "true" else "false"
            }
        }
        return null
    }

    fun setAttr(parentController: UIController, attr: String, value: String?) {
        if (value == null) {
            return
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            when (attr) {
                "RequestPermissions" -> {
                    val strs = value.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    // fnId=strs[0] , permissions=strs[1].split("|") , requestCode=strs[2]
                    val requestCode = Integer.valueOf(strs[2])
                    val ps = strs[1].split("#".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    parentController.onPermissionResults.put(requestCode, strs[0])
                    Log.d(TAG, "setAttr: $value")
                    ActivityCompat.requestPermissions(
                        parentController.activity,
                        ps,
                        requestCode
                    )
                }
            }
        }
    }
}
